Improved performance for writing large records (> blockSize)

amityahav opened the pull request:
#20 
improved performance for writing records which are bigger than the blockSize (32K), by using the one reused buffer, since writing is synchronized we can reuse chunkBuffer after each write. i've decided to initialize chunkBuffer with size of 32k and its capacity will grow as long as new larger records are coming in. this behavior is same with bytebuffer pool.
so there is a trade-off i see here between:

1. keep the same behavior which is write multiple times for large records but keep re-using a buffer with max capacity of blockSize.

OR

2.  optimize for a single i/o sys call but maintain a buffer that can grow indefinitely (and might cause OOM).

also removed redundant chunk slice allocations.

all tests passed and also added new benchmark 

before:
![image](https://github.com/rosedblabs/wal/assets/74831037/e9d2ff7d-6901-49c8-9278-c4eaa54b4e65)

after:
![image](https://github.com/rosedblabs/wal/assets/74831037/a4ea4aa0-f12f-4570-9ac5-8b773389fff5)

amityahav commented:
@roseduan

roseduan commented on segment.go:
```diff
@@ -259,27 +265,30 @@ func (seg *segment) writeInternal(data []byte, chunkType ChunkType) error {
-		seg.currentBlockSize = 0
+	seg.currentBlockSize += uint32(len(seg.chunkBuffer))
+
+	if seg.currentBlockSize >= blockSize {
```
Will seg.currentBlockSize be greater than blockSize?
I think they are equal at most.

amityahav replied:
@roseduan it can be greater than blockSize as I add the chunkBuffer size to it at line 282

amityahav commented on segment.go:
```diff
@@ -259,27 +265,30 @@ func (seg *segment) writeInternal(data []byte, chunkType ChunkType) error {
+
+	if seg.currentBlockSize >= blockSize {
+		sub := seg.currentBlockSize - blockSize
+		seg.currentBlockNumber += 1 + (sub / blockSize)
```
I will try to document/ simplify it . Anyhow can u check please the global description of the PR? I've written there some concerns regarding this improvement that I would like your opinion on. Thanks

roseduan commented:
> #20 improved performance for writing records which are bigger than the blockSize (32K), by using the one reused buffer, since writing is synchronized we can reuse chunkBuffer after each write. i've decided to initialize chunkBuffer with size of 32k and its capacity will grow as long as new larger records are coming in. this behavior is same with bytebuffer pool. so there is a trade-off i see here between:
> 
> 1. keep the same behavior which is write multiple times for large records but keep re-using a buffer with max capacity of blockSize.
> 
> OR
> 
> 2. optimize for a single i/o sys call but maintain a buffer that can grow indefinitely (and might cause OOM).
> 
> also removed redundant chunk slice allocations.
> 
> all tests passed and also added new benchmark
> 
> before: ![image](https://user-images.githubusercontent.com/74831037/259800702-e9d2ff7d-6901-49c8-9278-c4eaa54b4e65.png)
> 
> after: ![image](https://user-images.githubusercontent.com/74831037/259800291-a4ea4aa0-f12f-4570-9ac5-8b773389fff5.png)

I understand your consideration, I think we can use the bytebufferpool to cache all chunks and write it once into the file.
Then we can reset the buffer and put it back to the pool.

amityahav commented:
> > #20 improved performance for writing records which are bigger than the blockSize (32K), by using the one reused buffer, since writing is synchronized we can reuse chunkBuffer after each write. i've decided to initialize chunkBuffer with size of 32k and its capacity will grow as long as new larger records are coming in. this behavior is same with bytebuffer pool. so there is a trade-off i see here between:
> > 
> > 1. keep the same behavior which is write multiple times for large records but keep re-using a buffer with max capacity of blockSize.
> > 
> > OR
> > 
> > 2. optimize for a single i/o sys call but maintain a buffer that can grow indefinitely (and might cause OOM).
> > 
> > also removed redundant chunk slice allocations.
> > all tests passed and also added new benchmark
> > before: ![image](https://user-images.githubusercontent.com/74831037/259800702-e9d2ff7d-6901-49c8-9278-c4eaa54b4e65.png)
> > after: ![image](https://user-images.githubusercontent.com/74831037/259800291-a4ea4aa0-f12f-4570-9ac5-8b773389fff5.png)
> 
> I understand your consideration, I think we can use the bytebufferpool to cache all chunks and write it once into the file. Then we can reset the buffer and put it back to the pool.

great then its already implemented. ill fix the other rejects and push

amityahav commented:
@roseduan hey i've simplifyed and refactored abit
